coq-builtin.elpi

```diff
@@ -1802,9 +1802,20 @@ type export-module modpath -> synterp-action.
 % execution (aka interp). See also coq.replay-synterp-action
 external pred coq.next-synterp-action o:synterp-action.
 
-% [coq.replay-all-missing-synterp-actions] Execute all actions needed in
-% order to match the actions performed at parsing time (aka synterp)
-external pred coq.replay-all-missing-synterp-actions .
```

Comment:
Question, is this API still implementable?

Reply:
I mean, can we have a replay-missing-from-current-group and have a outermost group "all"?

Reply:
Yes, but it is not a great API since it pierces through abstractions, so we thought that it was better to remove it. The reason why we had to implement this new API is to avoid this precise problem in some of our complex elpi commands that rely on NES internally.

Reply:
Yes to your second question.

Reply:
Actually I don't know about group "all" in a context where you compose commands, it is not clear who would begin/end it. We can easily add a `replay-missing-from-current-group` though.

Reply:
Doesn't  coq.next-synterp-action pierces it? (it is just above)

Shoudn't you add two constructors to the synterp-action to be used when the next action is open/closing a group? In a way,coq.replay-all-missing-synterp-actions was also implementable via coq.next-synterp-action + an utility in coq-lib to replay one action (mentioned in the comment above).

So either you remove the coq.next-synterp-action (which I'm not very sure about) or make it return the open/closing bracket. IMO the latter would be a good compromise, since one can implement in pure elpi an a function that, for example, closes the current box, and go back using the high level APIs. I'm sure I use next-synterp-action in HB for example, not sure removing it is worth it.

Reply:
> Doesn't coq.next-synterp-action pierces it? (it is just above)

It fails if the group has not been opened, so no. At interp, you need to either use `replay-synterp-action-group "my-group"` or explicitly do `begin-synterp-group "my-group" G`, a bunch of `coq.next-synterp-action`, and `end-synterp-group G`. Potentially with more nesting if you have sub-groups.

Reply:
Here's what we could easily do in elpi itself: we made `coq.next-synterp-action` fail softly when there aren't any more actions so we could call that repeatedly until it fails and just replay the actions it returns. It would work at the top level just like the old primitive worked and it would also work to finish all actions in the current group but it wouldn't silently pierce anybody's group abstractions.

Reply:
What you propose seems sensible, let me just double check I understood.

It seems I could take any piece of code replaying some actions via the next-synterp-action and call it from a context that puts it in a box (a group) and the code should just work. Did I get it right? If so, 100% OK.

Reply:
Yes, that's exactly right.

Reply:
This is implemented now.